### PR TITLE
Suppress pub actions for Flutter projects.

### PR DESCRIPTION
**Pull request by pq**
Fixes: https://github.com/flutter/flutter-intellij/issues/660

I still need to do some testing (my setup to properly compile against the latest sources is currently broken) but would love any early feedback.

@alexander-doroshko

### CODE REVIEW

File: Dart/src/com/jetbrains/lang/dart/flutter/FlutterUtil.java [hunk 1]
```diff
@@ -8,27 +10,47 @@
...
+  }
+
+  /**
+   * Test if the given project defines a Flutter module.
+   *
+   * @param module the project to test
+   * @return true if the given project has a Flutter module, false otherwise.
+   */
+  public static boolean hasFlutterModule(@NotNull project) {
+    return Arrays.stream(ModuleManager.getInstance(project).getModules()).anyMatch(FlutterUtil::isFlutterModule);
```
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/flutter/FlutterUtil.java]: `ModuleUtil.hasModulesOfType` does it faster. You'll probably need to use `ModuleTypeManager.findByID` as well.

File: Dart/src/com/jetbrains/lang/dart/flutter/FlutterUtil.java [hunk 2]
```diff
@@ -8,27 +10,47 @@
...
    *
-   * @param dartSdk the Dart SDK
-   * @return the relative Flutter root, or null if this SDK is not relative to one
+   * @return true if the the Flutter plugin is installed, false otherwise.
    */
-  @Nullable
-  public static String getFlutterRoot(@NotNull final String dartSdkPath) {
-    final String suffix = "/bin/cache/dart-sdk";
-    return dartSdkPath.endsWith(suffix) ? dartSdkPath.substring(0, dartSdkPath.length() - suffix.length()) : null;
+  public static boolean isFlutterPluginInstalled() {
```
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/flutter/FlutterUtil.java]: this knowledge may be stored as a final static field, plugin list can't change without IDE restart.

File: Dart/src/com/jetbrains/lang/dart/ide/actions/DartEditorNotificationsProvider.java [hunk 3]
```diff
@@ -50,6 +51,12 @@ public EditorNotificationPanel createNotificationPanel(@NotNull final VirtualFil
     }
 
     if (PubspecYamlUtil.PUBSPEC_YAML.equalsIgnoreCase(vFile.getName())) {
+
+      // Defer to the Flutter plugin for package management and SDK configuration if appropriate.
+      if (FlutterUtil.isFlutterPluginInstalled() && FlutterUtil.hasFlutterModule(myProject)) {
+        return null;
```
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/ide/actions/DartEditorNotificationsProvider.java]: You disable notification wihtout checking current module type. Is this what you intended to do?

File: Dart/src/com/jetbrains/lang/dart/ide/actions/DartPubActionBase.java [hunk 4]
```diff
@@ -76,11 +76,19 @@ public static void setIsInProgress(boolean inProgress) {
...
-    final boolean visible = getModuleAndPubspecYamlFile(e) != null;
+    final boolean visible = getModuleAndPubspecYamlFile(e) != null && !shouldSuppressForFlutter(e);
     e.getPresentation().setVisible(visible);
     e.getPresentation().setEnabled(visible && !isInProgress());
   }
 
+  private static boolean shouldSuppressForFlutter(@NotNull final AnActionEvent e) {
+    if (FlutterUtil.isFlutterPluginInstalled()) {
+      final Project project = e.getProject();
+      return project != null && FlutterUtil.hasFlutterModule(project);
```
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/ide/actions/DartPubActionBase.java]: The same question: doesn't _current_ module type matter? There may be 2 modules in one project: one for Dart and one for Flutter, and this one may be a Dart one. Will flutter plugin handle it?

**Comment by alexander-doroshko** [comment]
Is this PR for discussion only? It is not compilable.

File: Dart/src/com/jetbrains/lang/dart/flutter/FlutterUtil.java (hunk 1 above)
Comment (pq) [inline Dart/src/com/jetbrains/lang/dart/flutter/FlutterUtil.java]: Looks like we won't need this after all (based on your comments below).

File: Dart/src/com/jetbrains/lang/dart/flutter/FlutterUtil.java (hunk 2 above)
Comment (pq) [inline Dart/src/com/jetbrains/lang/dart/flutter/FlutterUtil.java]: Good idea!

File: Dart/src/com/jetbrains/lang/dart/ide/actions/DartEditorNotificationsProvider.java (hunk 3 above)
Comment (pq) [inline Dart/src/com/jetbrains/lang/dart/ide/actions/DartEditorNotificationsProvider.java]: Right.  Nice catch.

File: Dart/src/com/jetbrains/lang/dart/ide/actions/DartPubActionBase.java (hunk 4 above)
Comment (pq) [inline Dart/src/com/jetbrains/lang/dart/ide/actions/DartPubActionBase.java]: 👍

**Comment by pq** [comment]
> Is this PR for discussion only? It is not compilable.

Sorry about that. I'm still struggling to get IDEA compiling again.  Whenever I come back to work on the Dart Plugin my environment needs to be re-setup. 😢

File: Dart/src/com/jetbrains/lang/dart/flutter/FlutterUtil.java (hunk 2 above)
Comment (pq) [inline Dart/src/com/jetbrains/lang/dart/flutter/FlutterUtil.java]: Fixed.

**Comment by pq** [comment]
More fixes.  Ready for another look @alexander-doroshko.

Thanks in advance!

**Comment by alexander-doroshko** [comment]
Merged manually, thanks!

**Comment by pq** [comment]
Thanks!

**Pull request closed without merging**